graphix/opengraph.py

```diff
@@ -0,0 +1,162 @@
+
+    from graphix.pattern import Pattern
+    from graphix.pauli import Plane
+
+
+class Measurement(NamedTuple):
+    """An MBQC measurement.
+
+    :param angle: the angle of the measurement. Should be between [0, 2)
+    :param plane: the measurement plane
+    """
+
+    angle: float
+    plane: Plane
+
+    def isclose(self, other: Measurement, rel_tol=1e-09, abs_tol=0.0) -> bool:
+        """Compares if two measurements have the same plane and their angles
+        are close.
+
+        Example
+        -------
+        >>> from graphix.opengraph import Measurement
+        >>> from graphix.pauli import Plane
+        >>> Measurement(0.0, Plane.XY).isclose(Measurement(0.0, Plane.XY))
+        True
+        >>> Measurement(0.0, Plane.XY).isclose(Measurement(0.0, Plane.YZ))
+        False
+        >>> Measurement(0.1, Plane.XY).isclose(Measurement(0.0, Plane.XY))
+        False
+        """
+        return math.isclose(self.angle, other.angle, rel_tol=rel_tol, abs_tol=abs_tol) and self.plane == other.plane
+
+
+class OpenGraph:
+    """Open graph contains the graph, measurement, and input and output
+    nodes. This is the graph we wish to implement deterministically
+
+    :param inside: the underlying graph state
+    :param measurements: a dictionary whose key is the ID of a node and the
+        value is the measurement at that node
+    :param inputs: a set of IDs of the nodes that are inputs to the graph
+    :param outputs: a set of IDs of the nodes that are outputs of the graph
+
+    Example
+    -------
+    >>> import networkx as nx
+    >>> from graphix.opengraph import OpenGraph, Measurement
+    >>>
+    >>> inside_graph = nx.Graph([(0, 1), (1, 2), (2, 0)])
+    >>>
+    >>> measurements = {i: Measurement(0.5 * i, Plane.XY) for i in range(2)}
+    >>> inputs = [0]
+    >>> outputs = [2]
+    >>> og = OpenGraph(inside_graph, measurements, inputs, outputs)
+    """
+
+    __inside: nx.Graph
+    __inputs: list[int]
+    __outputs: list[int]
+    __meas: Mapping[int, Measurement]
+
+    def __eq__(self, other) -> bool:
+        """Checks the two open graphs are equal
+
+        This doesn't check they are equal up to an isomorphism"""
+
+        if not nx.utils.graphs_equal(self._inside, other._inside):
+            return False
+
+        if self.inputs != other.inputs or self.outputs != other.outputs:
+            return False
+
+        if set(self.measurements.keys()) != set(other.measurements.keys()):
+            return False
+
+        return all(m.isclose(other.measurements[node]) for node, m in self.measurements.items())
+
+    def __init__(
+        self,
+        inside: nx.Graph,
+        measurements: Mapping[int, Measurement],
+        inputs: list[int],
+        outputs: list[int],
+    ) -> None:
+        """Constructs a new OpenGraph instance
+
+        The inputs() and outputs() methods will preserve the order that was
+        original given in to this methods.
+        """
+        self._inside = inside
```

Comment:
The field declared above is called `__inside` (private, with two `_`), not `_inside`. Same remark for `__inputs`, `__outputs` and `__meas`.

Reply:
Thanks for pointing this out, but I actually already changed these as per EarlMilkTea's suggestion in a previous commit. Not sure why the change didn't display on your side.

Reply:
Are you sure you pushed on the branch `opengraph` associated to this PR? The file `opengraph.py` on the last commit 5e7b2aa of this branch still contains occurrences of `_inline`: https://github.com/TeamGraphix/graphix/blob/5e7b2aab3c403b2da78dacfbad19b776600b42de/graphix/opengraph.py#L103

Reply:
@wlcsm I agree with @thierry-martinez, it seems like your changes are not on remote repo. I unresolved correpsonding @EarlMilktea 's somment. Could you try to push the changes again?

Reply:
Ah my apologies, my local repo wasn't correctly synchronised. It should be fixed now